sympy/polys/ring_series.py

```diff
@@ -229,8 +229,10 @@ def rs_mul(p1, p2, x, prec):
     """
     R = p1.ring
     p = R.zero
-    if R.__class__ != p2.ring.__class__ or R != p2.ring:
-        raise ValueError('p1 and p2 must have the same ring')
+    if R.__class__ != p2.ring.__class__:
+        raise TypeError("p1 and p2 must have the same ring class")
+    if R != p2.ring:
+        raise ValueError('p1 and p2 must have the same ring value')
```

Comment:
I don't understand what the purpose of this change is. Probably both cases should raise `TypeError` but I'm not sure it's worth changing that.